crates/db_schema/src/schema_setup.rs

```diff
@@ -0,0 +1,42 @@
+use anyhow::Context;
+use diesel::{connection::SimpleConnection, Connection, PgConnection};
+use diesel_migrations::{EmbeddedMigrations, MigrationHarness};
+use lemmy_utils::error::LemmyError;
+use tracing::info;
+
+const MIGRATIONS: EmbeddedMigrations = embed_migrations!();
+
+/// This SQL code sets up the `r` schema, which contains things that can be safely dropped and replaced
+/// instead of being changed using migrations. It may not create or modify things outside of the `r` schema
+/// (indicated by `r.` before the name), unless a comment says otherwise.
+///
+/// If you add something that depends on something (such as a table) created in a new migration, then down.sql
+/// must use `CASCADE` when dropping it. This doesn't need to be fixed in old migrations because the
+/// "replaceable-schema" migration runs `DROP SCHEMA IF EXISTS r CASCADE` in down.sql.
+const REPLACEABLE_SCHEMA: &[&str] = &[
+  "BEGIN;",
+  "DROP SCHEMA IF EXISTS r CASCADE;",
+  "CREATE SCHEMA r;",
+  include_str!("../replaceable_schema/utils.sql"),
+  include_str!("../replaceable_schema/triggers.sql"),
+  "COMMIT;",
+];
+
+pub fn run(db_url: &str) -> Result<(), LemmyError> {
+  // Migrations don't support async connection
+  let mut conn = PgConnection::establish(db_url).with_context(|| "Error connecting to database")?;
+
+  // Migrations
+  info!("Running Database migrations (This may take a long time)...");
+  conn
+    .run_pending_migrations(MIGRATIONS)
+    .map_err(|e| anyhow::anyhow!("Couldn't run DB Migrations: {e}"))?;
+  info!("Database migrations complete.");
+
+  // Replaceable schema
+  conn
+    .batch_execute(&REPLACEABLE_SCHEMA.join("\n"))
+    .context("Couldn't run SQL files in crates/db_schema/replaceable_schema")?;
```

Comment:
if I understand this right this will currently rerun all the triggers deletion + creation on every server start (regardless of whether the version changed or not). I don't think that's a great idea because even if the schema does not contain data dropping and rebuilding it is still going to do multiple unwanted things:

1. It will have to wait if there is any kind of lock on anything in the schema. For example, pg_dump takes a `ACCESS SHARE` lock. So if you have a pg_dump running, then with this change lemmy will hang on start up until the dump is done (also without showing any debuggable reasons why). Lots of things may take weak locks like this because they don't usually cause problems.
2. If someone is running a horizontally scaled setup (like most of the large instances do by now) they will expectedly run rolling restarts to minimize downtime when doing minor changes or auto-scaling or whatever other reason. During partial downtime, it seems like the schema will be dropped for a short time which means that any row updates happening during that time will not execute the triggers? During a version upgrade this _might_ be fine, but during regular server restart I don't think it is. The alternative of locking everything during this process isn't good either.


 I think the easiest way to fix this would be to only run the schema update if (a) running in dev mode and (b) the backend version has changed (might be possible to use that backend version id which is already detected somewhere in the rust code)

Reply:
I fixed it by making this only run if at least 1 pending migration was run.

In the future, this solution will probably be replaced by something that queries the contents of the "r" schema to determine what needs to be changed, because moving index declarations here (without redundantly rebuilding them) would require that.